### PR TITLE
Add VCR back for unit testing in InSpec

### PR DESCRIPTION
**Pull request by slevenick**
Adding back VCR testing for InSpec

Currently the InSpec test job does NOT block merging. This can be added once we have stability in the new InSpec test flow.

New process for adding or changing an InSpec resource test is:
1) Create PR with new or modified tests, inspec-test step will fail using cassettes from master
2) Run inspec-vcr-record step manually on the PR created via 1. This uploads new cassettes to a GS bucket for this specific PR
3) Kick off another build for the PR, this will pull the new cassettes created via 2, succeeding
4) Continue as normal

I don't currently have a way to update the master cassettes. Possible post merge-prs step? Possibly done overnight during nightly integration test runs though that will break the test step for the rest of the day after an InSpec PR is merged. Possible one off build that is run manually? I don't have a great solution yet

-----------------------------------------------------------------
# [all]

### CODE REVIEW

**Comment by slevenick** [comment]
I don't currently have a way to update the master cassettes. Possible post merge-prs step? Possibly done overnight during nightly integration test runs though that will break the test step for the rest of the day after an InSpec PR is merged. Possible one off build that is run manually? I don't have a great solution yet

File: .ci/ci.yml.tmpl [hunk 1]
```diff
@@ -276,7 +276,6 @@ jobs:
     - name: ansible-test
       plan:
         - get: magic-modules
-          version: every
```
Comment (rambleraptor) [inline .ci/ci.yml.tmpl]: Why is this being removed?

File: templates/inspec/singular_resource.erb [hunk 2]
```diff
@@ -15,13 +15,19 @@
 # frozen_string_literal: false
 
 <%= lines(autogen_notice :ruby) -%>
+<% vcr = ENV['VCR_MODE'] -%>
```
Comment (rambleraptor) [inline templates/inspec/singular_resource.erb]: Little confused by this.

It looks like we have two different sets of resources - one that work normally and one that only work with VCR cassettes. Is there a way of building VCR tests that work using the regular controls (so we don't have to build two versions that we assume work the same but may not)

File: templates/inspec/vcr_config.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  c.hook_into :webmock
+  c.cassette_library_dir = 'inspec-cassettes'
+  c.allow_http_connections_when_no_cassette = true
+
+  c.before_record do |i|
+    i.response.headers.delete_if { |key| key != 'Content-Type' }
+    i.request.headers.delete_if { |key| true }
+    if auth_call?(i)
+      i.request.body = 'AUTH REQUEST'
+      i.response.body = "{\n  \"access_token\": \"ya29.c.samsamsamsamsamsamsamsamsa-thisisnintysixcharactersoftexttolooklikeanauthtokenthisisnintysixcharactersoftexttolooklikeanaut\",\n  \"expires_in\": 3600,\n  \"token_type\": \"Bearer\"\n}"
```
Comment (rambleraptor) [inline templates/inspec/vcr_config.rb]: Could you do something like: `"ya29.c.#{'sam' * 30}"` to clean this up but still have a long string?

File: templates/inspec/singular_resource.erb [hunk 4]
```diff
@@ -15,13 +15,19 @@
...
 <%
   require 'google/string_utils'
 
+
   inside_indent = 8
 
   requires = generate_requires(object.all_user_properties)
   requires << 'gcp_backend'
+  if vcr
```
Comment (rambleraptor) [inline templates/inspec/singular_resource.erb]: `requires += ['vcr', 'vcr_config'] if vcr`

Comment (rambleraptor) [reply templates/inspec/singular_resource.erb]: (I'm pretty sure that'll work)

File: .ci/ci.yml.tmpl (hunk 1 above)
Comment (slevenick) [reply .ci/ci.yml.tmpl]: Whoops, typo

File: templates/inspec/vcr_config.rb (hunk 3 above)
Comment (slevenick) [reply templates/inspec/vcr_config.rb]: Yeah. The format of the access token is pretty specific, but I could totally clean it up.

**Comment by rambleraptor** [comment]
Quick recap of our conversation:

I wasn't a huge fan of generating two different sets of resources: one with VCR logic and one without because:

1) Testing logic living in non-test code (we don't want to have code for tests having side effects that affect real code. Test code should live separately from real code to make everyone's lives easier)
2) The resources we're running the VCR tests on are fundamentally different than the ones we're releasing into production. Yes, they're not different by much (a couple lines), but we would be running tests on different code than we would be releasing. That's worrying to me, especially as the code within the VCR blocks changes over time and we forget about this assumption.

At the same time, the InSpec folks don't want a lot of VCR logic within the tests because they view the tests as examples and tests.

Sam is going to try and build two sets of tests - one with VCR logic (will be tested, never released) and one without (to be released). While we run into the same issue as before (releasing code that's different than the code we test), I think it's a better idea to do this with tests.

File: templates/inspec/vcr_config.rb (hunk 3 above)
Comment (rambleraptor) [reply templates/inspec/vcr_config.rb]: This is such a nit, so your call on what to do if anything

File: templates/inspec/integration_test_template.erb [hunk 5]
```diff
@@ -1,4 +1,9 @@
 <%= lines(autogen_notice :ruby) -%>
+<% vcr = ENV['VCR_MODE'] -%>
```
Comment (rambleraptor) [inline templates/inspec/integration_test_template.erb]: I didn't realize that this ENV variable has expected values. Could you do some kind of check to throw an error if this var has a bad value?

**Comment by slevenick** [comment]
> Quick recap of our conversation:
> 
> I wasn't a huge fan of generating two different sets of resources one with VCR logic and one without) because:
> 
> 1. Testing logic would live in non-test code. We don't want to have code for tests having side effects that affect real code. Test code should live separately from real code to make everyone's lives easier and a separation between testing + not-testing should be the default.
> 2. The resources we would be running the VCR tests on are fundamentally different than the ones we would be releasing into production. Yes, they're not different by much (a couple lines), but we would be running tests on different code than we would be releasing. That's worrying to me, especially as the code within the VCR blocks changes over time and we forget about this assumption.
> 
> At the same time, the InSpec folks don't want a lot of VCR logic within the tests because they view the tests as examples and tests.
> 
> Sam is going to try and build two sets of tests - one with VCR logic (will be tested, never released) and one without (to be released). While we run into the same issue as before (releasing code that's different than the code we test), I think it's a better idea to do this with tests. Both sets of tests would still be testing the exact same code.

Should be fixed. Moved vcr config to be in tests. Ran into issues around caching and dependencies on resources that are not MM generated, but fixed all of the current ones.

**Comment by ndmckinley** [comment]
> I don't currently have a way to update the master cassettes. Possible post merge-prs step? Possibly done overnight during nightly integration test runs though that will break the test step for the rest of the day after an InSpec PR is merged. Possible one off build that is run manually? I don't have a great solution yet

I suggest that your first approach is the best one - outside of merge-prs, triggers on successful merge, generate new cassettes for master.

**Pull request merged**